documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc

```diff
@@ -2093,14 +2095,60 @@ public class VehicleMapperImpl extends VehicleMapper {
 ----
 ====
 
-Only methods with return type `void` may be annotated with `@BeforeMapping` or `@AfterMapping`. The methods may or may not have parameters.
-
 If the `@BeforeMapping` / `@AfterMapping` method has parameters, the method invocation is only generated if all parameters can be *assigned* by the source or target parameters of the mapping method:
 
 * A parameter annotated with `@MappingTarget` is populated with the target instance of the mapping.
 * A parameter annotated with `@TargetType` is populated with the target type of the mapping.
 * Any other parameter is populated with a source parameter of the mapping, whereas each source parameter is used once at most.
 
+If the before/after-mapping method has a return type other than `void`, it will be checked to match the target type of the mapping methods.
+Only the callback methods with a matching return type (or `void`) will be called in that mapping method. 
+
+If a callback method returns a non-null value, this value will be returned from the mapping method.
+
+As with mapping methods, it is possible to specify type parameters for before/after-mapping methods.
+
+.Mapper with @AfterMapping hook that returns a non-null value
+====
+[source, java, linenums]
+[subs="verbatim,attributes"]
+----
+@Mapper
+public abstract class VehicleMapper {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @AfterMapping
+    protected <T> T attachEntity(@MappingTarget Object entity) {
```

Comment:
@pascalgn If the `@MappingTarget` is an `Object` it still does not work. In the `MethodMatcher#matchResultType` and `MethodMatcher#matchSourceType` we have some explicit check to see that the the candidate result/source type is not `java.lang.Object`.

Can you change your the example to:

```java
@AfterMapping
public static <T> T attachObject(@MappingTarget T object) {
    return entityManager.merge(object);
}
```
 This would be similar to the test